testgen/src/light_block.rs

```diff
@@ -0,0 +1,273 @@
+///         Self {
+///             signed_header: tm_lb.signed_header,
+///             validators: tm_lb.validators,
+///             next_validators: tm_lb.next_validators,
+///             provider: tm_lb.provider,
+///         }
+///     }
+/// }
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub struct TMLightBlock {
+    /// Header and commit of this block
+    pub signed_header: SignedHeader,
+    /// Validator set at the block height
+    pub validators: ValidatorSet,
+    /// Validator set at the next block height
+    pub next_validators: ValidatorSet,
+    /// The peer ID of the node that provided this block
+    pub provider: PeerId,
+}
+
+/// We use this data structure as a simplistic representation of LightClient's LightBlock
+#[derive(Debug, Options, Deserialize, Clone)]
+pub struct LightBlock {
+    #[options(help = "header (required)", parse(try_from_str = "parse_as::<Header>"))]
+    pub header: Option<Header>,
+    #[options(help = "commit (required)", parse(try_from_str = "parse_as::<Commit>"))]
+    pub commit: Option<Commit>,
+    #[options(
+        help = "validators (required), encoded as array of 'validator' parameters",
+        parse(try_from_str = "parse_as::<Vec<Validator>>")
+    )]
+    pub validators: Option<Vec<Validator>>,
+    #[options(
+        help = "next validators (default: same as validators), encoded as array of 'validator' parameters",
+        parse(try_from_str = "parse_as::<Vec<Validator>>")
+    )]
+    pub next_validators: Option<Vec<Validator>>,
+    #[options(help = "peer id (default: default_peer_id())")]
+    pub provider: Option<PeerId>,
+}
+
+impl LightBlock {
+    /// Constructs a new Testgen-specific light block
+    pub fn new(header: Header, commit: Commit) -> Self {
+        Self {
+            header: Some(header),
+            commit: Some(commit),
+            validators: None,
+            next_validators: None,
+            provider: None,
+        }
+    }
+    set_option!(validators, &[Validator], Some(validators.to_vec()));
+    set_option!(
+        next_validators,
+        &[Validator],
+        Some(next_validators.to_vec())
+    );
+
+    pub fn new_default(height: u64) -> Self {
+        let validators = [Validator::new("1"), Validator::new("2")];
+        let header = Header::new(&validators)
+            .height(height)
+            .chain_id("test-chain")
+            .next_validators(&validators)
+            .time(height); // just wanted to initialize time with some value
+
+        let commit = Commit::new(header.clone(), 1);
+
+        Self {
+            header: Some(header),
+            commit: Some(commit),
+            validators: Some(validators.to_vec()),
+            next_validators: Some(validators.to_vec()),
+            provider: Some(default_peer_id()),
+        }
+    }
+
+    /// Produces a subsequent, i.e. at (height+1), light block to the supplied one
+    // TODO: figure how to represent the currently ignored details in header
```

Comment:
Left a few TODOs here but with the current state you should be able to use it for testing so long as you don't care about the ignored details.

Reply:
Indeed, this should not pose a problem at the moment.